app/channels/job_outputs_channel.rb

```diff
@@ -50,33 +53,49 @@ def status_response(event)
   # some kind of buffered broadcast channel would be ideal
   def subscribed
     id = params.fetch(:id)
-    if execution = JobQueue.find_by_id(id)
-      execution.viewers.push current_user
-      Thread.new do
-        ActiveRecord::Base.connection_pool.with_connection do
-          builder = EventBuilder.new(execution.job)
-          execution.output.each do |event, data|
-            transmit event: event, data: builder.payload(event, data)
-          end
-          # TODO: disconnect all listeners so they close their sockets ?
-          # then replace the reloaded/finished/waitUntilEnabled stuff with that
+
+    @thread = Thread.new do
+      ActiveRecord::Base.connection_pool.with_connection do
+        (MAX_WAIT / WAIT_DURATION).times do
+          # running
+          execution = JobQueue.find_by_id(id)
+          break stream_execution(execution) if execution
+
+          # cancelled or we missed it running
+          job = Job.find(id)
+          break stream_finished_output(job) if job.finished?
+
+          # pending
+          sleep WAIT_DURATION
         end
```

Comment:
does it close the connection if it gets here (times out waiting)?

Reply:
the channel does not get closed automatically, but since it's not doing anything it does not really matter (same as before)

Reply:
seems like it should get closed if it gets to here, unless there's a reason not to (like it'll reconnect or something)

Reply:
if we do a `connection.close` then the client reconnects and streams the whole output again :)
... there might be a way to make this work, but it involves a lot of digging/checking for little gain (closing connections that do nothing, not sure if it even removes the channel and which are auto-closed when user closes the tab)